src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,101 +3,356 @@
+        // without worrying about this blocking the reading thread on cross-process pipe io.
+        var fillReadResult = await _pipeReader.ReadAtLeastAsync(length, cancellationToken).ConfigureAwait(false);
+
+        // Note: we have let the pipe reader know that we're done with 'read at least' call, but that we
+        // haven't consumed anything from it yet.  Otherwise it will throw that another read can't start
+        // from within ObjectReader.GetReader below.
+        _pipeReader.AdvanceTo(fillReadResult.Buffer.Start);
+
+        // Let the object reader do it's own individual object checking.
+        objectReader.CheckValidationBytes();
+
+        // Now do the actual read of the data, synchronously, from the buffers that are now in memory within our
+        // process.  These reads will move the pipe-reader forward, without causing any blocking on async-io.
+        var checksum = Checksum.ReadFrom(objectReader);
+        var kind = (WellKnownSynchronizationKind)objectReader.ReadByte();
+
+        var asset = _serializer.Deserialize(kind, objectReader, cancellationToken);
+        Contract.ThrowIfNull(asset);
+        _callback(checksum, (T)asset, _arg);
+    }
+
+    private async ValueTask<int> CheckSentinelByteAndReadLengthAsync(CancellationToken cancellationToken)
+    {
+        const int HeaderSize = sizeof(byte) + sizeof(int);
+
+        var lengthReadResult = await _pipeReader.ReadAtLeastAsync(HeaderSize, cancellationToken).ConfigureAwait(false);
+        var (sentinelByte, length) = ReadSentinelAndLength(lengthReadResult);
+
+        // Check that the sentinel is correct, and move the pipe reader forward to the end of the header.
+        Contract.ThrowIfTrue(sentinelByte != RemoteHostAssetWriter.MessageSentinelByte);
+        _pipeReader.AdvanceTo(lengthReadResult.Buffer.GetPosition(HeaderSize));
+
+        return length;
+    }
+
+    // Note on Checksum itself as it depends on SequenceReader, which is provided by nerdbank.streams on
+    // netstandard2.0 (which the Workspace layer does not depend on).
+    private async ValueTask<Checksum> ReadChecksumFromPipeReaderAsync(CancellationToken cancellationToken)
+    {
+        var readChecksumResult = await _pipeReader.ReadAtLeastAsync(Checksum.HashSize, cancellationToken).ConfigureAwait(false);
+
+        var checksum = ReadChecksum(readChecksumResult);
+        _pipeReader.AdvanceTo(readChecksumResult.Buffer.GetPosition(Checksum.HashSize));
+        return checksum;
+    }
+
+    private static (byte, int) ReadSentinelAndLength(ReadResult readResult)
+    {
+        var sequenceReader = new SequenceReader<byte>(readResult.Buffer);
+        Contract.ThrowIfFalse(sequenceReader.TryRead(out var sentinel));
+        Contract.ThrowIfFalse(sequenceReader.TryReadLittleEndian(out int length));
+        return (sentinel, length);
+    }
+
+    private static Checksum ReadChecksum(ReadResult readResult)
```

Comment:
>ReadChecksum [](http://example.com/codeflow?start=28&length=12)

nit: this method and it's only caller are both only 4 lines, so, might be less interruptive to just inline this method in the caller.

Reply:
i can't.  they use spans.  which don't work in async methods :)   we're fixing that, but not until c# 13.